### PR TITLE
Short term archive

### PR DESCRIPTION
**Pull request by mvertens**
New pythonized short term archiving capability. This permits creating rpointer files on the fly for interim restart files.

Test suite: scripts_regression_tests
Test baseline: NA
Test namelist changes: none
Test status: [bit for bit, roundoff, climate changing] All passed except for ERI_D.f45_g37.X.yellowstone_intel - but this is not failing in the short term archiving.

Fixes: [CIME Github issue #]

User interface changes?: 
config_archive.xml (used to be archive.xml) has a new xml schema - but this is not a user interface change.

Code review: J
im Edwards, Alice Bertini

### CODE REVIEW

File: utils/python/CIME/XML/env_archive.py [hunk 1]
```diff
@@ -17,87 +17,73 @@ def __init__(self, case_root=None, infile="env_archive.xml"):
...
+    def get_rpointer_contents(self, archive_entry):
+        rpointer_items = []
+        rpointer_nodes = self.get_nodes('rpointer', root=archive_entry)
+        for rpointer_node in rpointer_nodes:
+            file_node = self.get_node('rpointer_file', root=rpointer_node)
+            content_node = self.get_node('rpointer_content', root=rpointer_node)
+            rpointer_items.append([file_node.text,content_node.text])
+        return rpointer_items
+
+    # def get_value(self, item, attribute={}, resolved=True, subgroup=None):
```
Comment (jgfouca) [inline utils/python/CIME/XML/env_archive.py]: Can this commented-out stuff be removed?

File: utils/python/CIME/case.py [hunk 2]
```diff
@@ -378,6 +378,27 @@ def _get_component_config_data(self):
             if result is not None:
                 del self.lookups[key]
 
+    def get_components(self):
+        # return dictionary of the form [component_class:component],
```
Comment (jgfouca) [inline utils/python/CIME/case.py]: These first two comments should be in docstring form:

    """
    doc
    """

File: utils/python/CIME/case_st_archive.py [hunk 3]
```diff
@@ -1,24 +1,413 @@
...
+
+    logger.debug('In check_run...')
+    dout_s_root = case.get_value('DOUT_S_ROOT')
+    if dout_s_root is None or dout_s_root == 'UNSET':
+        expect(False, 'XML variable DOUT_S_ROOT is required for root location of short-term achiver')
+    if not isdir(dout_s_root):
+        os.makedirs(dout_s_root)
+    dout_s_save_interim = case.get_value('DOUT_S_SAVE_INTERIM_RESTART_FILES')
+    if dout_s_save_interim == 'FALSE' or dout_s_save_interim == 'UNSET':
+        rest_n = int(case.get_value('REST_N'))
```
Comment (jgfouca) [inline utils/python/CIME/case_st_archive.py]: These options are known to be integer types so will be returned as ints. There's no need for the casts

File: utils/python/CIME/case_st_archive.py [hunk 4]
```diff
@@ -1,24 +1,413 @@
...
+                    if not ninst_strings:
+                        ninst_strings = ["empty"]
+                for ninst_string in ninst_strings:
+                    if ninst_string == 'empty':
+                        ninst_string = ""
+                    subs = dict()
+                    subs['$CASE'] = casename
+                    subs['$DATENAME'] = datename
+                    subs['$NINST_STRING'] = ninst_string
+                    for key in subs.keys():
```
Comment (jgfouca) [inline utils/python/CIME/case_st_archive.py]: for key, value in [ ('$CASE', casename), ('$DATENAME', datename), ('$NINST_STRING', ninst_string) ]:

File: utils/python/CIME/case_st_archive.py [hunk 5]
```diff
@@ -1,24 +1,413 @@
...
+        cmd = "ncdump -v %s %s " %(rest_hist_varname, os.path.join(rundir,restfile))
+        rc,out,error = run_cmd(cmd, ok_to_fail=True)
+        searchname = "%s =" %rest_hist_varname
+        if searchname in out:
+            offset = out.index(searchname)
+            items  = out[offset:].split(",")
+            for item in items:
+                # the following match has an option of having a './' at the beginning of
+                # the history filename
+                matchobj = re.search("\"(\.*\/*\w.*)\s?\"",item)
```
Comment (jgfouca) [inline utils/python/CIME/case_st_archive.py]: This needs to be a raw string. Did you run this through pylint?

File: utils/python/CIME/case_st_archive.py [hunk 6]
```diff
@@ -1,24 +1,413 @@
...
+def get_histfiles_for_restarts(case, archive, archive_entry, restfile):
+###############################################################################
+
+    # determine history files that are needed for restarts 
+    histfiles = []
+    rest_hist_varname = archive.get_entry_value('rest_history_varname', archive_entry)
+    if rest_hist_varname != 'unset':
+        rundir = case.get_value("RUNDIR")
+        cmd = "ncdump -v %s %s " %(rest_hist_varname, os.path.join(rundir,restfile))
+        rc,out,error = run_cmd(cmd, ok_to_fail=True)
```
Comment (jgfouca) [inline utils/python/CIME/case_st_archive.py]: It looks like you ignored the return code. Should we be checking that it succeeded?

**Comment by jgfouca** [comment]
@mvertens : Looks good. Just had a few comments and I'd like to have you run this through pylint.

File: utils/python/CIME/XML/env_archive.py (hunk 1 above)
Comment (mvertens) [inline utils/python/CIME/XML/env_archive.py]: yes. Done.

File: utils/python/CIME/case.py (hunk 2 above)
Comment (mvertens) [inline utils/python/CIME/case.py]: done

File: utils/python/CIME/case_st_archive.py (hunk 3 above)
Comment (mvertens) [inline utils/python/CIME/case_st_archive.py]: done

File: utils/python/CIME/case_st_archive.py (hunk 4 above)
Comment (mvertens) [inline utils/python/CIME/case_st_archive.py]: thanks. this is much cleaner. done.

File: utils/python/CIME/case_st_archive.py (hunk 6 above)
Comment (mvertens) [inline utils/python/CIME/case_st_archive.py]: thanks for catching this. done.

**Comment by mvertens** [comment]
Ran pylint and fixed lots of things. I'm running the regression test again and will hopefully have an updated PR today.

**Comment by mvertens** [comment]
Just updated the PR based on successful run of the regression tests.

**Comment by jgfouca** [comment]
@mvertens ready for me to merge?

**Comment by mvertens** [comment]
yes

**Pull request merged**